horovod/common/ops/cuda/adasum_cuda_kernels.cu

```diff
@@ -0,0 +1,199 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <iostream>
+#include <stdio.h>
+#include <cuda_fp16.h>
+#include <time.h>
+#include <stdint.h>
+
+#define THREADS_PER_BLOCK 64
+
+template<typename T, typename TACC>
+__global__
+void CudaSingleAdasumKernel(int count, T* a, const T* b, TACC* out) {
+	__shared__ TACC a_coeff;
+	__shared__ TACC b_coeff;
+	int index = threadIdx.x + blockIdx.x * blockDim.x;
+	if (0 == threadIdx.x) {
+		if ((TACC) out[0] != 0) {
+			a_coeff = 1.0 - out[2] / out[0] * 0.5;
+		} else {
+			a_coeff = 1;
+		}
+		if ((TACC) out[1] != 0) {
+			b_coeff = 1.0 - out[2] / out[1] * 0.5;
+		} else {
+			b_coeff = 1;
+		}
+	}
+	__syncthreads();
+	if (count > index){
+		a[index] = (T) ((TACC) a[index] * a_coeff + (TACC) b[index] * b_coeff);
+	}
+}
+
+template<typename T, typename TACC>
+__global__
+void CudaDotProductKernel(int count, const T* a, const T* b, TACC* out) {
+	__shared__ TACC normsq_a[THREADS_PER_BLOCK];
+	__shared__ TACC normsq_b[THREADS_PER_BLOCK];
+	__shared__ TACC dot[THREADS_PER_BLOCK];
+	int index = threadIdx.x + blockIdx.x * blockDim.x;
+	if (index < count){
+		normsq_a[threadIdx.x] = (TACC) a[index] * (TACC) a[index];
+		normsq_b[threadIdx.x] = (TACC) b[index] * (TACC) b[index];
+		dot[threadIdx.x]      = (TACC) a[index] * (TACC) b[index];
+	}
+	__syncthreads();
+	if (0 == threadIdx.x) {
+		TACC normsq_a_sum = 0;
+		TACC normsq_b_sum = 0;
+		TACC dot_sum = 0;
+		for(int i = 0; i < THREADS_PER_BLOCK; i++){
+			if (i + blockIdx.x * blockDim.x < count){
+				normsq_a_sum += normsq_a[i];
+				normsq_b_sum += normsq_b[i];
+				dot_sum += dot[i];
+			}
+		}
+		atomicAdd(out, normsq_a_sum);
+		atomicAdd(out+1, normsq_b_sum);
+		atomicAdd(out+2, dot_sum);
+	}
+}
+
+template<typename T, typename TACC>
+__global__
+void CudaScaleAddKernel(int count, T* a, const T* b, TACC a_coeff, TACC b_coeff) {
+	int index = threadIdx.x + blockIdx.x * blockDim.x;
+	if (count > index){
+		a[index] = (T) ((TACC) a[index] * a_coeff + (TACC) b[index] * b_coeff);
+	}
+}
+
+template<typename T>
+__global__
+void ConvertToFloat(int count, T* a, float* b) {
+	int index = threadIdx.x + blockIdx.x * blockDim.x;
+	if (count > index){
+		b[index] = (float) a[index];
+	}
+}
+
+void ErrorCheck(std::string op_name, cudaError_t cuda_result) {
+	if (cuda_result != cudaSuccess) {
+		throw std::logic_error(std::string(op_name) + " failed: " + cudaGetErrorString(cuda_result));
+	}
+}
+
+void CudaSingleAdasumImpl(int count, double* device_a, const double* device_b, 
+	double* device_vals, cudaStream_t stream) {
+	
+	ErrorCheck("cudaMemsetAsync", cudaMemsetAsync(device_vals, 0, 3*sizeof(double), stream));
+
+	CudaDotProductKernel<<<(count+THREADS_PER_BLOCK-1)/THREADS_PER_BLOCK,
```

Comment:
i think you can even fuse this and the adasumSingleKernel into one kernel or call adasumSingleKernel from DotProductKernel and make adasumSingleKernel __device__ instead of __global__. That will save some time for the host kernel launch.

Reply:
The issue is that I need a synchronization between the two, wherein AdasumSingleKernel requires all of the DotProductKernel to have finished executing. I could not figure out a way to do this while merging them into a single kernel. Not sure how calling one kernel from inside another works though, would that force global device sync?

Reply:
yea so if you just change CudaSingleAdasumKernel from __global__ to __device__ and call it from CudaDotProductKernel, these 2 kernels will be fused into one big kernel. And synchronization within one kernel is ensured.

Reply:
So, that is not a bad idea, but you will need a global barrier which new cuda allows you to do:
https://devblogs.nvidia.com/cooperative-groups/

The question is, is it worth it? Do you know how the performance changes by removing one kernel? (I know the result would be bogus but just for performance measuring purposes).